media-proxy/src/mesh/conn_rdma_rx.cc

```diff
@@ -0,0 +1,149 @@
+                              (" ",kind_to_string(_kind));
+                    add_to_queue(buf);
+                }
+            } else {
+                break; // Exit the loop to avoid spinning on errors
+            }
+        }
+        // Wait for the buffer to become available
+        wait_buf_available();
+    }
+}
+
+/**
+ * @brief Handles the RDMA completion queue (CQ) events in a dedicated thread.
+ * 
+ * This function continuously monitors the CQ for completion events, processes
+ * completed buffers, and returns them to the buffer queue for reuse. It ensures
+ * efficient event handling by reading events in batches and sleeping briefly
+ * when no events are available to avoid busy waiting.
+ * 
+ * Key Steps:
+ * 1. Reads a batch of CQ entries.
+ * 2. For each entry:
+ *    - Processes the received buffer and sends it.
+ *    - Adds the buffer back to the queue upon successful processing.
+ *    - Logs errors for any issues during processing or queue operations.
+ * 3. Handles `EAGAIN` errors by yielding CPU time to avoid busy looping.
+ * 4. Logs and exits on other CQ read errors.
+ * 
+ * @param ctx The context for managing thread cancellation and operations.
+ */
+void RdmaRx::rdma_cq_thread(context::Context& ctx)
+{
+    constexpr int CQ_RETRY_DELAY_US = 100; // Retry delay for EAGAIN
+    struct fi_cq_entry cq_entries[CQ_BATCH_SIZE]; // Array to hold batch of CQ entries
+
+    while (!ctx.cancelled()) {
+        // Read a batch of completion events
+        int ret = fi_cq_read(ep_ctx->cq_ctx.cq, cq_entries, CQ_BATCH_SIZE);
+        if (ret > 0) {
+            for (int i = 0; i < ret; ++i) {
+                void *buf = cq_entries[i].op_context;
+                if (buf == nullptr) {
+                    log::error("Null buffer context, skipping...")
+                              ("batch_index",i)(" ", kind_to_string(_kind));
+                    continue;
+                }
+
+                // Process the buffer (e.g., transmit or handle)
+                Result res = transmit(ctx, buf, trx_sz);
+                if (res != Result::success) {
+                    log::error("Failed to transmit buffer")("buffer_address", buf)("size", trx_sz)
+                              (" ", kind_to_string(_kind));
+                    continue;
+                }
+
+                // Add the buffer back to the queue
+                res = add_to_queue(buf);
+                if (res == Result::success) {
+                    // Notify that a buffer is available
+                    notify_buf_available();
+                } else {
+                    log::error("Failed to add buffer back to the queue")("buffer_address", buf)
+                              (" ", kind_to_string(_kind));
+                }
+            }
+        } else if (ret == -EAGAIN) {
+            // No events to process, yield CPU briefly to avoid busy looping
+            std::this_thread::sleep_for(std::chrono::microseconds(CQ_RETRY_DELAY_US));
```

Comment:
Add new function inside concurrency.cc
```
void Sleep(context::Context& ctx, std::chrono::microseconds interval_us)
{
    std::mutex mx;
    std::unique_lock<std::mutex> lk(mx);
    std::condition_variable_any cv;
    cv.wait_for(lk, ctx.stop_token(), interval_us, [] { return false; });
}
```
And replace usage of ```this_thread::sleep``` with:

Reply:
Well, I don't think we need a delay of less than a millisecond to be cancellable.
@moleksy Keep it as it is for microsecond delays.

Reply:
Monitoring if the context has been cancelled is needed only on operations blocking for 100ms or more. In other cases it is safe to use `this_thread::sleep_for`.

Reply:
Sound reasonable